src/stan/interface_callbacks/writer/base_writer.hpp

```diff
@@ -8,15 +9,55 @@ namespace stan {
   namespace interface_callbacks {
     namespace writer {
 
+      // base_writer is an absract base class defining the
+      // interface for Stan writer callbacks.  The Stan API
+      // writes structured output to implementations of this
+      // class defined by a given interface.
+      //
+      // A typical API function signature will be of the form
+      //
+      // template <class WriterA, class WriterB>
+      // void function(WriterA& writer_a, WriterB& writer_b) {
+      //   ...
+      //   writer_a(key, value);
+      //   writer_b("message");
+      //   ...
+      // }
+      //
+      // where WriterA and WriterB are base_writer implementations.
+
       class base_writer {
       public:
+        // Virtual destructor to avoid compiler warnings
         virtual ~base_writer() {}
-        // Can't enforce this method with a pure virtual function
-        // template <class T>
-        // virtual void operator()(const std::vector<T>& x) = 0;
-        virtual void operator()(const std::string x) = 0;
+
+        // Key-scalar input
+        virtual void operator()(const std::string& key,
```

Comment:
@bob-carpenter, think we should name these methods? The idea that this is a functor is nice, but I think the use here is a bit confusing. I don't have a suggestion for naming, though.